### PR TITLE
Drive Base

### LINKED ISSUE #78: Choose between multiple autonomous modes
Should be able to easily change our autonomous code to work with a different alliance, amend issues, etc.

### PR DESCRIPTION
**Pull request by tonyn10**
External encoders and auton modes/command groups

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -21,11 +21,11 @@
...
-    CDSSubsystem      (true),
-    IntakeSubsystem   (true),
-    ShooterSubsystem  (true),
-    LimelightSubsystem(true),
-    ClimbSubsystem    (true);
+    CDSSubsystem      (false),
+    IntakeSubsystem   (false),
+    ShooterSubsystem  (false),
+    LimelightSubsystem(false),
+    ClimbSubsystem    (false);
```
Comment (sarango2158) [inline src/main/java/frc/robot/Constants.java]: reset to true before merging this PR into main

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -213,43 +206,22 @@ private void configureButtonBindings() {
...
-                  Constants.ramseteZeta), // ramsete follower to follow trajectory
-              Constants.driveKinematics,
-              driveBaseSubsystem::acceptWheelSpeeds,
-              driveBaseSubsystem);
-
-      driveBaseSubsystem.resetOdometry(trajectory.getInitialPose());
-
-      return ramseteCommand.andThen(() -> driveBaseSubsystem.acceptWheelSpeeds(0, 0));
-    }
+    return chosenAutonMode;
```
Comment (sarango2158) [inline src/main/java/frc/robot/RobotContainer.java]: I think this still needs to return null if drive base is disabled. Possibly have other subsystems that we should check now too

File: src/main/java/frc/robot/commands/AutonModes.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            new WaitCommand(Constants.delayshot),
+            // new ShooterPrime(shooterSubsystem, limelightSubsystem, cdsSubsystem),
+            new ShooterPressed(
+                shooterSubsystem,
+                limelightSubsystem,
+                cdsSubsystem,
+                true), // also has a time delay of 2-3 seconds
+            new WaitCommand(Constants.delaytaxi),
+            oneBallRamseteCommands[0].beforeStarting(
+                () -> driveBaseSubsystem.resetOdometry(taxiTrajectories[0].getInitialPose())));
```
Comment (azvanderpas) [inline src/main/java/frc/robot/commands/AutonModes.java]: taxiTrajectories should be oneBallTrajectories

File: src/main/java/frc/robot/commands/AutonModes.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            // new ShooterPrime(shooterSubsystem, limelightSubsystem, cdsSubsystem)
+            new ShooterPressed(shooterSubsystem, limelightSubsystem, cdsSubsystem, true));
+  }
+
+  private Trajectory[] getTrajectories(String routeString) {
+    // method to intialize multiple trajectories in one auton route
+    String routeName = routeString; // TODO: replace route for the correct one on pathweaver
+    File f =
+        new File(
+            "Autos/" + routeName); // get file with all the names of the path for the auton routine
```
Comment (azvanderpas) [inline src/main/java/frc/robot/commands/AutonModes.java]: Remember that only files/folders in the deploy directory will get copied to the roboRIO. You need to move Autos to the deploy directory.

File: src/main/java/frc/robot/commands/AutonModes.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    Scanner fileScanner = null;
```
Comment (azvanderpas) [inline src/main/java/frc/robot/commands/AutonModes.java]: No need to assign to null
```suggestion
    Scanner fileScanner;
```

**Review by sarango2158 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/AutonModes.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    twoBallParallel =
+        new ParallelDeadlineGroup(
+            twoBallRamseteCommands[0].beforeStarting(
+                () -> driveBaseSubsystem.resetOdometry(twoBallTrajectories[0].getInitialPose())),
+            new IntakeForwardCommand(intakeSubsystem));
+    twoBallCommand =
+        new SequentialCommandGroup(
+            new WaitCommand(Constants.delaytaxi),
+            twoBallParallel,
+            twoBallRamseteCommands[1], // goes to fender
```
Comment (azvanderpas) [inline src/main/java/frc/robot/commands/AutonModes.java]: Does the odometry need to be reset here as well?

File: src/main/java/frc/robot/commands/AutonModes.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+      ramseteCommands[i] =
+          new RamseteCommand(
+              trajectories[i],
+              driveBaseSubsystem::getPose,
+              new RamseteController(
+                  Constants.ramseteB,
+                  Constants.ramseteZeta), // ramsete follower to follow trajectory
+              Constants.driveKinematics,
+              driveBaseSubsystem::acceptWheelSpeeds,
+              driveBaseSubsystem);
```
Comment (azvanderpas) [inline src/main/java/frc/robot/commands/AutonModes.java]:
```suggestion
              driveBaseSubsystem).beforeStarting(
                  () -> driveBaseSubsystem.resetOdometry(trajectories[i].getInitialPose()));
```

Then its bundled into all the ramsete commands and you don't have to redo it for each in initializeCommandGroups

File: src/main/java/frc/robot/commands/AutonModes.java [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+      Path path =
+          Filesystem.getDeployDirectory()
+              .toPath()
+              .resolve(pathName); // goes to scr/main/deploy/paths
+      Trajectory trajectory;
+      try {
+        trajectory = TrajectoryUtil.fromPathweaverJson(path);
+        trajectoryList.add(trajectory);
+      } catch (IOException e) {
+        System.out.println("Trajectory: " + pathName + " not created.");
```
Comment (caseyjbrotherton) [inline src/main/java/frc/robot/commands/AutonModes.java]: We should consider whether we want to have a driverstation error for this:

https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/wpilibj/DriverStation.html#reportError(java.lang.String,boolean)

Not sure.  
Pro: Easier to see.  This error is important to know if one of our auton modes are broke.
Con:  More complicated, and we shouldn't see this error often, if at all.

**Review by caseyjbrotherton — APPROVED** [review]

File: src/main/java/frc/robot/Constants.java [hunk 9]
```diff
@@ -108,30 +108,34 @@ public AimModes previous() {
...
 
   // drive base pid values
   public static final double[] driveRightPID = {0.00035, 0.0000008, 0};
   public static final double[] driveLeftPID = {0.000005, 0.0000008, 0};
 
   // AUTONOMOUS Constants
-
+  public static final String taxiout = "TaxiOut.wpilib.json";
+  public static final String taxioutfender = "TaxiOutFromFender.wpilib.json";
+  public static final String taxioutball = "TaxiOutGrabBall.wpilib.json";
```
Comment (azvanderpas) [inline src/main/java/frc/robot/Constants.java:138]: How about an array?

File: src/main/java/frc/robot/Constants.java [hunk 10]
```diff
@@ -107,35 +107,43 @@ public AimModes previous() {
...
-  public static final int driveLeftFront = 6;
-  public static final int driveLeftRear = 7;
-  public static final int driveRightFront = 13;
-  public static final int driveRightRear = 14;
+
+  // TODO: kit bot values for now, change later
+  public static final int driveLeftFront = 1; // 13 on real robot
+  public static final int driveLeftRear = 2; // 14 on real robot
+  public static final int driveRightFront = 3; // 6 on real robot
+  public static final int driveRightRear = 4; // 7 on real robot
```
Comment (azvanderpas) [inline src/main/java/frc/robot/Constants.java]: Revert before merging

**Review by azvanderpas — APPROVED** [review]

**Review by sarango2158 — APPROVED** [review]

**Pull request merged**